app/main.py

```diff
@@ -1,2 +1,159 @@
+from fractions import Fraction
+from typing import Any, Hashable, Iterable, Optional
+
+
 class Dictionary:
-    pass
+    class __Node:
+        def __init__(
+                self,
+                key: Optional[Hashable] = None,
+                value: Optional[Any] = None,
+                key_hash: Optional[int] = None
+        ) -> None:
+            self.key = key
+            self.value = value
+            self.key_hash = key_hash
+            self.deleted = False
+
+    def __init__(
+            self,
+            initial_data: Optional[Iterable[tuple[Hashable, Any]]] = None,
+            **kwargs
+    ) -> None:
+        self.__capacity = 8
+        self.__size = 0
+        self.__threshold = int(Fraction(2, 3) * self.__capacity)
+        self.__table = [Dictionary.__Node()] * self.__capacity
+
+        if initial_data:
+            self.update(initial_data)
+        if kwargs:
+            self.update(**kwargs)
+
+    def __repr__(self) -> str:
+        items = ", ".join(f"{key}: {value}" for key, value in self.items())
+        return f"{{{items}}}"
+
+    def __setitem__(self, key: Hashable, value: Any) -> None:
+        index = self.__calculate_index(key)
+
+        if self.__table[index].key is None:
+            if self.__size + 1 > self.__threshold:
+                self.__resize()
+                return self.__setitem__(key, value)
+            self.__size += 1
+
+        self.__table[index] = Dictionary.__Node(key, value, hash(key))
+
+    def __getitem__(self, key: Hashable) -> Any:
+        index = self.__calculate_index(key)
+
+        if self.__table[index].key is None:
+            raise KeyError(key)
+
+        return self.__table[index].value
+
+    def __delitem__(self, key: Hashable) -> None:
+        index = self.__calculate_index(key)
+
+        if self.__table[index].key is None:
+            raise KeyError(key)
+
+        self.__table[index] = Dictionary.__Node()
+        self.__table[index].deleted = True
+        self.__size -= 1
+
+    def __len__(self) -> int:
+        return self.__size
+
+    def __iter__(self) -> "Dictionary":
+        self.__current_index = 0
+        self.__keys = (node.key for node in self.__table if node.key)
```

Comment:
you can simply return this generator expression and it will be most efficient iterator. Then you will not need to store __keys and implement __next__ method

Reply:
Done